In [2]:
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [3]:
%load_ext sql

In [4]:
%sql postgresql://postgres:postgres@localhost:5430/eskom_db

In [5]:
%%sql 
DROP TABLE IF EXISTS eskom_data;
CREATE TABLE  eskom_data(
    date_time_hour_beginning DATE,
    original_res_forecast_before_lockdown NUMERIC,
    residual_forecast  NUMERIC,	
    rsa_contracted_forecast	 NUMERIC,
    dispatchable_generation	NUMERIC ,
    residual_demand NUMERIC,
    rsa_contracted_demand  NUMERIC,
    international_exports  NUMERIC,
    international_imports  NUMERIC,
    thermal_generation  NUMERIC,
    nuclear_generation  NUMERIC,
    eskom_gas_generation  NUMERIC,	
    eskom_ocgt_generation  NUMERIC,	
    hydro_water_generation  NUMERIC,
    pumped_water_generation	 NUMERIC,
    ils_usage  NUMERIC,
    manual_load_reduction NUMERIC,
    ios_excl_ils_and_mlr  NUMERIC,
    dispatchable_ipp_ocgt  NUMERIC,
    eskom_gas_sco  NUMERIC,
    eskom_ocgt_sco  NUMERIC,
    hydro_water_sco	 NUMERIC,
    pumped_water_sco  NUMERIC,
    pumping	 NUMERIC,
    wind  NUMERIC,
    pv  NUMERIC,
    csp  NUMERIC,
    other_re  NUMERIC,	
    total_re  NUMERIC,	
    wind_installed_capacity	 NUMERIC,
    pv_installed_capacity  NUMERIC,	
    csp_installed_capacity  NUMERIC,
    other_re_installed_capacity	 NUMERIC,
    total_re_installed_capacity	 NUMERIC,
    installed_eskom_capacity  NUMERIC,
    total_pclf  NUMERIC,
    total_uclf  NUMERIC,
    total_oclf  NUMERIC,
    total_uclf_and_oclf	 NUMERIC,
    non_comm_sentout  NUMERIC,	
    drakensberg_gen_unit_hours  NUMERIC,
    palmiet_gen_unit_hours  NUMERIC,	
    ingula_gen_unit_hours NUMERIC
    
);


 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
Done.


[]

In [6]:
%%sql 
COPY eskom_data(date_time_hour_beginning,original_res_forecast_before_lockdown,residual_forecast,rsa_contracted_forecast,dispatchable_generation,residual_demand,rsa_contracted_demand,international_exports,international_imports,thermal_generation,nuclear_generation,eskom_gas_generation,eskom_ocgt_generation,hydro_water_generation,pumped_water_generation,ils_usage,manual_load_reduction,ios_excl_ils_and_mlr,dispatchable_ipp_ocgt,eskom_gas_sco,eskom_ocgt_sco,hydro_water_sco,pumped_water_sco,pumping,wind,pv,csp,other_re,total_re,wind_installed_capacity,pv_installed_capacity,csp_installed_capacity,other_re_installed_capacity,total_re_installed_capacity,installed_eskom_capacity,total_pclf,total_uclf,total_oclf,total_uclf_and_oclf,non_comm_sentout,drakensberg_gen_unit_hours,palmiet_gen_unit_hours,ingula_gen_unit_hours)
FROM '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/ESK2033.csv'
DELIMITER ','
CSV HEADER
;

 * postgresql://postgres:***@localhost:5430/eskom_db
37704 rows affected.


[]

In [7]:
%%sql 
DROP TABLE IF EXISTS generation,demand,capacity;
CREATE TABLE demand
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    date_time_hour_beginning,
    residual_demand,
    rsa_contracted_demand
FROM eskom_data;
CREATE TABLE generation
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    dispatchable_generation,
    international_exports,
    international_imports,
    thermal_generation,
    nuclear_generation,
    eskom_gas_generation,	
    eskom_ocgt_generation,	
    hydro_water_generation,
    pumped_water_generation,
    dispatchable_ipp_ocgt,
    pumping,
    wind,
    pv,
    csp,
    other_re,	
    total_re,	
    drakensberg_gen_unit_hours,
    palmiet_gen_unit_hours,	
    ingula_gen_unit_hours
FROM eskom_data;
CREATE TABLE capacity
AS 
SELECT
    EXTRACT(YEAR FROM date_time_hour_beginning) as year,
    EXTRACT(MONTH FROM date_time_hour_beginning) as month,
    EXTRACT(DAY FROM date_time_hour_beginning) as day,
    wind_installed_capacity,
    pv_installed_capacity,	
    csp_installed_capacity,
    other_re_installed_capacity,
    total_re_installed_capacity,
    installed_eskom_capacity
FROM eskom_data;
COPY generation,demand,capacity TO '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/ESK2033_clean.csv' WITH DELIMITER ',' CSV HEADER;



 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
37704 rows affected.
37704 rows affected.
37704 rows affected.
(psycopg2.errors.SyntaxError) syntax error at or near ","
LINE 1: COPY generation,demand,capacity TO '/Users/ds_learner16/Docu...
                       ^

[SQL: COPY generation,demand,capacity TO '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/ESK2033_clean.csv' WITH DELIMITER ',' CSV HEADER;]
(Background on this error at: https://sqlalche.me/e/14/f405)


In [8]:
%%sql
DROP TABLE IF EXISTS Loadshedding;
CREATE TABLE loadshedding (
    created_at TIMESTAMP,
    stage INT
      
);

 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
Done.


[]

In [9]:
%%sql
COPY Loadshedding(created_at,stage)
FROM '/Users/ds_learner16/Documents/Eskom/Eskom_Data_Analysis/EskomSePush_history.csv'
DELIMITER ','
CSV HEADER;

 * postgresql://postgres:***@localhost:5430/eskom_db
670 rows affected.


[]

In [10]:
%%sql 
DROP TABLE IF EXISTS load_shedding_clean;
CREATE TABLE load_shedding_clean
AS
SELECT 
    to_char(created_at,'Mon') as month_name,
    to_char(created_at,'Day') as day_name,
    EXTRACT(YEAR FROM created_at) as year,
    EXTRACT(MONTH FROM created_at) as month,
    EXTRACT(HOUR FROM created_at) as hour,
    EXTRACT(isodow FROM created_at) as day,
    stage

FROM Loadshedding;



 * postgresql://postgres:***@localhost:5430/eskom_db
Done.
670 rows affected.


[]

# Number of Rows and Columns 
count rows = 42824 estimated before clean after 37704
removed forecast dates added average for original res before lockdown

count cols = 41

## Which year was the demand high

In [11]:
high_demand = %sql SELECT year as "year" , AVG(residual_demand) as "residual_demand", AVG(rsa_contracted_demand) as "rsa_contracted_demand" FROM demand GROUP BY year ORDER BY year
high_demand

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('25767.131784696970'), Decimal('26998.603180303030')),
 (Decimal('2019'), Decimal('25219.673938127854'), Decimal('26542.384553082192')),
 (Decimal('2020'), Decimal('23696.602886042805'), Decimal('25117.220372153916')),
 (Decimal('2021'), Decimal('24196.092115525114'), Decimal('25932.202556392694')),
 (Decimal('2022'), Decimal('24394.402998125000'), Decimal('26094.779150833333'))]

In [12]:
fig = px.line(high_demand,x="year",y=["residual_demand","rsa_contracted_demand"],title="Residual vs RSA Contracted Demand")
fig.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig.show()

## Which of the renewable energy generation methods have the highest capacity

In [13]:
renew_capacity = %sql SELECT year as "year",AVG(wind_installed_capacity) as "wind",AVG(pv_installed_capacity) as "pv",AVG(csp_installed_capacity) as "csp",AVG(other_re_installed_capacity) as "other" FROM capacity GROUP BY year ORDER BY year
renew_capacity

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[(Decimal('2018'), Decimal('1474.1900000000000000'), Decimal('311.6363636363636364'), Decimal('21.7800000000000000'), Decimal('3886.6987272727272727')),
 (Decimal('2019'), Decimal('1474.1900000000000000'), Decimal('491.7808219178082192'), Decimal('21.7800000000000000'), Decimal('4067.5108219178082192')),
 (Decimal('2020'), Decimal('1780.3017486338797814'), Decimal('500.0000000000000000'), Decimal('21.7800000000000000'), Decimal('4439.6280874316939891')),
 (Decimal('2021'), Decimal('2194.0434246575342466'), Decimal('500.0000000000000000'), Decimal('25.0073972602739726'), Decimal('5380.9399726027397260')),
 (Decimal('2022'), Decimal('2212.0900000000000000'), Decimal('500.0000000000000000'), Decimal('41.8300000000000000'), Decimal('5893.4680000000000000'))]

In [14]:
fig1 = px.line(renew_capacity,x="year",y=["wind","pv","csp","other"],title="Average of renewable energy capacity")
fig1.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig1.show()

## Which Year , month , hour had the most loadshedding 

In [32]:
yearly = %sql SELECT year as "year" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5",count(case when stage = 6 then 1 else null end) as "stage_6"  FROM load_shedding_clean GROUP BY year ORDER by year
yearly
fig2 = px.line(yearly,x="year",y=["stage_1","stage_2","stage_3","stage_4","stage_5","stage_6"],title="Count of stage by year")
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
6 rows affected.


In [31]:
monthly = %sql SELECT month,month_name as "month_name" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5",count(case when stage = 6 then 1 else null end) as "stage_6"  FROM load_shedding_clean GROUP BY month,month_name ORDER by month
monthly
fig2 = px.line(monthly,x="month_name",y=["stage_1","stage_2","stage_3","stage_4","stage_5","stage_6"],title="Count of stage by month")
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
12 rows affected.


In [27]:
dayly = %sql SELECT day,day_name as "day_name" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5",count(case when stage = 6 then 1 else null end) as "stage_6" FROM load_shedding_clean GROUP BY day,day_name ORDER by day
dayly
fig2 = px.line(dayly,x="day_name",y=["stage_1","stage_2","stage_3","stage_4","stage_5","stage_6"],title="Count of stage by day")
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
7 rows affected.


In [25]:
hourly = %sql SELECT hour as "hour" ,count(case when stage = 1 then 1 else null end) as "stage_1",count(case when stage = 2 then 1 else null end) as "stage_2",count(case when stage = 3 then 1 else null end) as "stage_3",count(case when stage = 4 then 1 else null end) as "stage_4",count(case when stage = 5 then 1 else null end) as "stage_5",count(case when stage = 6 then 1 else null end) as "stage_6" FROM load_shedding_clean GROUP BY hour ORDER by hour
hourly
fig2 = px.line(hourly,x="hour",y=["stage_1","stage_2","stage_3","stage_4","stage_5","stage_6"],title="Count of stage by Hour")
fig2.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig2.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
21 rows affected.


## Average energy that is generated yearly by eskom  in the country

In [19]:
yearly_dispatch = %sql select year as "year",AVG(dispatchable_generation) as "Dispatchable_generation(GW)" FROM generation GROUP BY year ORDER BY year
yearly_dispatch
fig3 = px.line(yearly_dispatch,x="year",y="Dispatchable_generation(GW)",title="Average of Dispatchable energy generation")
fig3.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig3.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


## Do we import more energy or export more energy?

In [20]:
imp_vs_exp = %sql select year as "year",avg(international_imports) as "imports", avg(international_exports) as "exports" from generation group by year order by year
imp_vs_exp
fig4 = px.line(imp_vs_exp,x="year",y=["imports","exports"],title="Imports vs Exports")
fig4.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig4.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


## Which one of the energy generation methods yeild the most power in GW

In [21]:
gen_method = %sql select year as "year",AVG(thermal_generation) as "thermal",AVG(nuclear_generation) as "nuclear",AVG(eskom_gas_generation) as "eskom_gas",AVG(eskom_ocgt_generation) as "eskom_ocgt",AVG(pumped_water_generation) as "pumped water",AVG(wind) as "wind",AVG(pv) as "pv",AVG(csp) as "csp",AVG(other_re) as "other_re" , avg(dispatchable_ipp_ocgt) as "ipp_ocgt" FROM generation GROUP BY year ORDER BY year
gen_method
fig5 = px.line(gen_method,x="year",y=["eskom_gas","eskom_ocgt","pumped water","nuclear","wind","pv","csp","other_re","ipp_ocgt"],title="Average of energy generation from different sources")
fig5.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig5.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


## Do we rely more on thermal energy or other generation methods

In [22]:
therm_vs_other = %sql select year as "year", avg(thermal_generation) as "thermal",avg(nuclear_generation+eskom_gas_generation+eskom_ocgt_generation+pumped_water_generation+wind+pv+csp+other_re+total_re+dispatchable_ipp_ocgt) as "other" from generation group by year order by year
fig6 = px.line(therm_vs_other,x="year",y=["thermal","other"],title="Thermal energy vs other energy")
fig6.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig6.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


## Which is the highest performing power station out of the 3 stations

In [23]:
high_perfom_stat = %sql select year as "year" , avg(drakensberg_gen_unit_hours) as "Drakensburg", avg(palmiet_gen_unit_hours) as "Palmiet",avg(ingula_gen_unit_hours) as "Ingula" from generation group by year order by year 
high_perfom_stat
fig7 = px.line(high_perfom_stat, x="year",y=["Drakensburg","Palmiet","Ingula"],title="Average of unit hours generated per power station")
fig7.update_layout(
    xaxis= dict(
        dtick = 1
    )
)
fig7.show()

 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


 * postgresql://postgres:***@localhost:5430/eskom_db
5 rows affected.


[('Jul', 'Monday   ', Decimal('2022'), Decimal('7'), Decimal('16'), Decimal('1'), 6),
 ('Jun', 'Thursday ', Decimal('2022'), Decimal('6'), Decimal('14'), Decimal('4'), 6),
 ('Jun', 'Wednesday', Decimal('2022'), Decimal('6'), Decimal('16'), Decimal('3'), 6),
 ('Jun', 'Tuesday  ', Decimal('2022'), Decimal('6'), Decimal('16'), Decimal('2'), 6),
 ('Dec', 'Monday   ', Decimal('2019'), Decimal('12'), Decimal('18'), Decimal('1'), 6)]